LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceRegistry.cs

```diff
@@ -362,6 +334,7 @@ public void ResetDeviceCache()
         {
             var oldResetCacheToken = this.resetCacheToken;
             this.resetCacheToken = new CancellationTokenSource();
+            this.resetCacheChangeToken = new CancellationChangeToken(this.resetCacheToken.Token);
```

Comment:
I think you need to lock this. Otherwise you can do this wil adding items to the cache and invalidating them right away.

Reply:
Do you mean adding a lock to prevent invalidating items that are being added as the cache is cleared?

If that is the case, I think it is better to use lock and let items that were just added to cache be invalidated instead of adding a lock before adding anything to the cache. Cache clear is supposed to be called not too often. If we start to rely on it more heavily we should be able to invalidate only parts of the cache.

Reply:
Not sure, but I think we mean the same thing: 
What I find dangerous here is that while calling ResetDeviceCache there might be entries coming in. Some of them might get removed and invalidated (added after the new resetCacheToken was initialized) and others might be removed (those that still got the original token).